core/src/main/scala/zio/logging/LogAppender.scala

```diff
@@ -57,21 +55,49 @@ object LogAppender {
           putStrLn(msg)
     ).map(appender => Has(appender.get.filter((ctx, _) => ctx.get(LogAnnotation.Level) >= logLevel)))
 
-  def file[A: Tag](filename: String, format0: LogFormat[A]) =
-    ZManaged.makeEffect {
-      val path = Paths.get(filename)
-      Files.newBufferedWriter(path)
-    }(_.close())
+  def file[A: Tag](destination: Path, format0: LogFormat[A]): ZLayer[Any, Throwable, Appender[A]] =
+    ZManaged
+      .makeEffect(Files.newBufferedWriter(destination))(_.close())
       .map(writer =>
         new Service[A] {
-
           override def write(ctx: LogContext, msg: => A): UIO[Unit] =
-            ZIO.effectTotal(writer.write(format0.format(ctx, msg)))
+            ZIO.effectTotal {
+              writer.write(s"${format0.format(ctx, msg)}${System.lineSeparator}")
+            }
         }
       )
       .toLayer
 
-  def ignore[A: Tag] =
+  def fileAsync[A: Tag](
```

Comment:
I was thinking that `async` should be a functionality that you can add on top of any appender.

```
LogAppender.file(...) >>> LogAppender.async
```
and 
```
LogAppender.console(...) >>> LogAppender.async
```

Let me know if you can implement it this way.

Reply:
Hmm, I agree it would be cool to be able to change any logger to be async easily, but I'm not seeing how it can be done generically like that considering the implementations can differ wildly between sync and async. For example, `file` and `fileAsync` don't even use the same type of Writer.

However, I do see the point that there's an opportunity to factor out much of the logic in `fileAsync` into some sort of abstraction that can be shared. At least half the code in `fileAsync` would be common logic.

Unless you have some other ideas that I'm not considering?

Reply:
That does work to a certain extent, but my concern was about how the implementations need to differ if you want a true performance gain. For example, I only want to flush a batch, not for every `write` call.

The difference is more noticeable in the file appender, but it's true even for the console appender. The default println implementation is `synchronized` and flushes for every line.

I'm not currently seeing a way to accomplish this generically. At least not with the current `Service` trait.

Reply:
what about using `BufferedWriter` like that `new BufferedWriter(new FileWriter(...), sizeOfBufferHere)` ?

Reply:
The rabbit hole is quite deep. There are a bunch of things going on:
- The ZQueue implementation is working off the number of "log entries" while BufferedWriter is working at the level of bytes.
- Setting the BufferedWriter size can lead to slightly surprising behavior because FileWriter (or more accurately StreamEncoder) also has its own internal buffer size. Maybe not really noticeable if you choose a larger buffer size though.
- Not flushing manually and working with number of bytes is a possible alternative, but you can get into a state where 1 log entry is only partially written to a file (with the rest still in a buffer). This might not be a problem, but worth noting I think.
- All these Writers are needlessly synchronized. I don't think that's a major concern, but I wonder if there's an NIO solution or something that's more "ideal" for this kind of usage.

I'm fine with a "good enough" solution for a first implementation and then refining it later some point in the future. I understand not wanting to force a big refactor out of this issue. Or not wanting to having 2 separate implementations for each appender.

Reply:
Writing our own small internal `Writer` (mostly just a wrapper) that is "log entry aware" for use in both sync and async might work? I don't know. Just a thought.

Reply:
I was looking for some NIO stuff but it seems that there is nothing out of the box that will help us. 

I like the idea of creating wrapper writer that will be lines aware. Let try this.

Reply:
we may also count number of the logs and flush when reach limit or periodically flush